web3.js/src/transaction.js

```diff
@@ -256,19 +254,33 @@ export class Transaction {
       }
     });
 
-    this.signatures.forEach(signature => {
+    // Move payer to the front and append other unknown signers as read-only
+    this.signatures.forEach((signature, signatureIndex) => {
       const sigPubkeyString = signature.publicKey.toString();
+      const isPayer = signatureIndex === 0;
```

Comment:
Does this assumption hold? For instance the first time `compileMessage()` is called, `this.signatures` hasn't yet been populated.  There also doesn't seem to be a path to an explicit payer, who isn't referenced by any instructions

Reply:
@t-nelson The way I view `compileMessage` is that it should only be called when the signers for a transaction have already been determined. This is because the compiled message is only useful for signing / verifying the transaction info that will be sent and this info won't be correct if signers aren't known upfront.

How about we remove the `this.signatures` population from `compileMessage` and require `this.signatures` to at least indicate what public keys should be signers. This would force devs to indicate all signers upfront with `signPartial` if they want to compile the message for offline signing. I think this forcing function would be a good thing because it would help prevent bugs where a dev compiles a message before specifying the payer, has the message signed by the payer, and then hits a signature verification error.

Thoughts?

Reply:
Thinking more on this... requiring a list of signers upfront feels redundant if a dev already specified certain accounts as `isSigner`. Perhaps only the fee payer needs to be made explicit and specified upfront.

There is a subtle bug that's still possible:

Imagine an instruction lists an account meta as read-only but then signs the transaction with that address. This alters the message header and would cause signature verification issues. We could try to handle this case or just throw an error if we manage to detect it.

Reply:
@t-nelson I've updated the PR to reflect our discussion, this PR is now more about requiring a fee payer to be specified. As a side effect, we are able to ensure that it is the first account meta. Let me know what you think!

Reply:
I still feel a little uneasy about the reliance on `self.signatures[0]` storing the payer _before_ `compileMessage()` is called.  It adds enough corner cases that I'm not confident I'm catching them all.  How about adding a `feePayer` member to transaction and storing it there (clearing `self.signatures` every time it changes), then inject it once into `accountMetas` in `compileMessage()`.  That's easier for me to reason about anyway.  WDYT?

Reply:
@t-nelson what kind of corner cases are you uneasy about? I feel that adding a separate `feePayer` field would make state more difficult to manage. It naturally fits in as the first signature.

I think `compileMessage` should never be called before expected signers are specified. Otherwise the message header could change. Having `compileMessage` have mutable behavior also felt like an anti-pattern.